plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -704,6 +716,91 @@ class PersonOverrideWriter {
     }
 }
 
+export class DeferredPersonOverrideWriter {
+    /**
+     * @param lockId the lock identifier/key used to ensure that only one
+     *               process is updating the overrides at a time
+     */
+    constructor(private postgres: PostgresRouter, private lockId: number) {}
+
+    /**
+     * Enqueue an override for deferred processing.
+     */
+    public async addPersonOverride(tx: TransactionClient, mergeOperation: MergeOperation): Promise<ProducerRecord[]> {
+        await this.postgres.query(
+            tx,
+            SQL`
+            INSERT INTO posthog_pendingpersonoverride (
+                team_id,
+                old_person_id,
+                override_person_id,
+                oldest_event
+            ) VALUES (
+                ${mergeOperation.team_id},
+                ${mergeOperation.old_person_id},
+                ${mergeOperation.override_person_id},
+                ${mergeOperation.oldest_event}
+            )`,
+            undefined,
+            'pendingPersonOverride'
+        )
+
+        return []
+    }
+
+    /**
+     * Process all pending overrides. An advisory lock is acquired prior to
+     * processing to ensure that this function has exclusive access to the
+     * pending overrides during the update process.
+     *
+     * @returns the number of overrides processed
+     */
+    public async processPendingOverrides(kafkaProducer: KafkaProducerWrapper): Promise<number> {
+        const writer = new PersonOverrideWriter(this.postgres)
+
+        return await this.postgres.transaction(PostgresUse.COMMON_WRITE, 'processPendingOverrides', async (tx) => {
+            const {
+                rows: [{ acquired }],
+            } = await this.postgres.query(
+                tx,
+                SQL`SELECT pg_try_advisory_xact_lock(${this.lockId}) as acquired`,
+                undefined,
+                'processPendingOverrides'
+            )
+            if (!acquired) {
+                throw new Error('could not acquire lock')
+            }
+
+            // n.b.: Ordering by id ensures we are processing in (roughly) FIFO order
```

Comment:
I could have sworn we discussed this and figured it all out but I'll have to look through our chats. 😰 

How are you thinking we _could_ check for transitive updates on both sides?

I wonder if this is why we had (for a moment, anyway) discussed how this `"pending"` table could instead be the one and only source of truth, rather than rows of a queue that come and go. Like, I think we had discussed a table that was something like `(team_id, old_person_id, override_person_id, shortened_path_person_id, oldest_event)`. So `shortened_path_person_id` would be updated as needed, and it would of course need indexes in different places if we wanted to do transitive lookups.

Unless you have figured out the solution here maybe we should discuss outside of comments here again and write down our findings.

Reply:
I was mostly worried before about the potential to create cycles in the merge history by foregoing the exclusion constraint (simplest example being A <- B and B <- A showing up in the pending person overrides), but I don't _think_ that is possible due to [the person deletion here](https://github.com/PostHog/posthog/blob/becc5127ec4e322aa13814542bf887764e2ef3af/plugin-server/src/worker/ingestion/person-state.ts#L516): we'd fail to get the lock [when updating the person when attempting to merge in the other direction](https://github.com/PostHog/posthog/blob/becc5127ec4e322aa13814542bf887764e2ef3af/plugin-server/src/worker/ingestion/person-state.ts#L495-L503).[^1] [This block also helps](https://github.com/PostHog/posthog/blob/becc5127ec4e322aa13814542bf887764e2ef3af/plugin-server/src/worker/ingestion/person-state.ts#L453-L457) (assuming `created_at` is immutable, and has a strict total ordering, i.e. no values are equal — which this probably isn't, but is probably close enough in practice), since that should keep us from incidentally creating a cycle between multiple persons because any sequence of merges within a set of users will always just end up pointing to the oldest user in the set. 

That is the stuff that I recall talking about before, but there may have been more beyond that stuff that has fallen out of my brain already. The order of operations stuff is a bit different.

Thinking out loud a bit here (there are probably a bunch of graph theoretic terms that would make these explanations clearer — or at least more precise — bear with me):

Generalizing this as a graph, I think that as long as there aren't any cycles in it (the merge graph), we should at least in theory be able to add new edges to it (i.e. append person merges) in basically any order.

The person overrides table isn't really a representation of the graph (though pending person overrides would be, if we never deleted from it — it's essentially an adjacency list), and is instead an index of the shortest path between an arbitrary node and whatever node it points to that only has edges pointing _at_ it and none away from it (i.e. only shows up in the merge graph as an override person, never as an old person.)

If we are processing events in order, A <- B <- C will only ever arrive as B <- C first, A <- B second. B <- C can't come after A <- B, because B would have been deleted. This makes it easier to keep our shortest path index, because once B <- C happens, we can just update all records for C to reference B and never consider it again, i.e. all records that had an override of C now can have an override of B. C <- D isn't a valid new entry, though it may have been an old one, updated in the previous step that now reflects that B <- D.

If we are processing events out of order, we lose this advantage. We can see _either_ A <- B _or_ B <- first. If the operations come _in_ order, we just do what was written above (and what the code already does.) If the operations come _out_ of order, (A <- B, then B <- C) we need to update the B <- C event as part of the insertion operation to reflect the fact that we've already seen A <- B _and do this prior to_ updating any records that already pointed at C to now point to A like before. Since we don't know if operations are arriving out of order, we need to treat them as if they were all the time. We don't need to do anything recursively because we're not dealing with a graph with intermediate edges (by design.)

I'm not sure how the squash operation would play into this yet. If we were to squash in between two out of order operations (e.g. A <- B, squash, B <- C), I think that means we could have some dangling merges that don't ever "complete" unless we keep a copy of the Postgres overrides around forever to identify the disconnect.

--

[^1]: Aside: I guess that maybe these locks should be taken explicitly if they matter that much? Also, there is a caveat here that the person UUID can't be reused again, otherwise it could appear in the merge history later and cause a cycle — but since it's a time-based UUID, I really hope that's not a problem.****

Reply:
My 2 🪙  on how to think of the ordering problem

So if we think about what ingestion does, then it has a single transaction that merges the person and adds the info to `posthog_pendingpersonoverride` table. In order to make this more clear I'll use capital letters to denote distinctIds that are being merged and lower case letters to denote personIDs that distinct_id maps to:
```
B(b) <- C(c)   ==> B(b) & C(b)  
A(a) <- B(b)   ==> A(a) & B(a) & C(a)
```
when we think of person overrides we'd get:
```
b <- c   ... applying transitive updates we get .... a <- c
a <- b
```
Now the other way (note how the second row has B(a) because person b doesn't exist anymore, which means we result in a <- c in the person overrides):
```
A(a) <- B(b)  ==> A(a) & B(a)
B(a) <- C(c)  ==> A(a) & B(a) & C(a)
```
when we think of person overrides we'd get:
```
a <- b
a <- c
```

Does this help?

For the squash part we discussed that if we cut off at a random point we'll just end up doing more updates, no problem ... e.g. if we squash, when we have `b <- c` but don't yet know that `a <- b` and hence `a <- c` if we do squash on top of `b <-c` we'd overwrite events to have person_id = `b` that then combined with the overrides table (`a <- b`) tells us that all those events are of person_id = `a`. So it works fine.

Reply:
> Does this help?

I think that's a good explanation of the correct/expected behavior. It's also a helpful reminder that the person overrides are a derivative of the distinct IDs, as well.

The override writer right now only handles the first case.

> For the squash part we discussed that if we cut off at a random point we'll just end up doing more updates, no problem ... e.g. if we squash, when we have b <- c but don't yet know that a <- b and hence a <- c if we do squash on top of b <-c we'd overwrite events to have person_id = b that then combined with the overrides table (a <- b) tells us that all those events are of person_id = a. So it works fine.

That example makes sense -- but in that example, the overrides are visible in order. If they arrive out of order, we'd end up in this scenario:

1. pending override for A <- B becomes visible and we update overrides to reflect A <- B
2. (squash happens, [overrides are deleted](https://github.com/PostHog/posthog/blob/14d3db3157f46f51ed5c5e34cf1c52152d336e0b/posthog/temporal/batch_exports/squash_person_overrides.py#L504))
3. pending override for B <- C becomes visible (out of order for some reason) to reflect B <- C (because we no longer have knowledge of A <- B to update the transitive override)

Don't all of the events that were previously associated with C get "lost" in that case, because they no longer reference a user that exists?

I'm not sure there's a good way around that, but here are some options I can think of:

1. ensure all inserts to the pending person overrides actually become visible in primary key order: either by locking the table as part of the merge transaction (yuck) or possibly by using serializable isolation level in the override worker and dealing with the potential for needing to retry due to serialization failures (also not great)
2. not deleting overrides ever (simple, but unbounded growth is not ideal)
3. define some transaction grace period, where we only select rows as inputs to squash that are older than X minutes, and assume that all transactions prior within that window have settled and that all valid insertions are now visible (maybe the best option?)

Reply:
That's not possible. We use postgres & transactions. There's no way we'd be able to write them out of order. Which ever distinctID merge happens first sets the person overrides in place for that. Then the later happening transaction would operate on the personIDs that are now mapping to those distinctIDs. You can't see a 
`b <- a` after a `c <- b`, because the latter would have combined b&c, so the lookup for b would have returned c.

Reply:
I'm going to merge this as-is for now and deal with these issues as a follow-up: I think the potential for pending overrides becoming visible out-of-order is ultimately low probability but high consequence, and I suspect we're very likely doing enough row locking here that the merge transaction that rows that affect the same persons forced into a serial execution path anyway (but I'm not sure how to prove that -- and even if I could, I think that'd be a tenuous guarantee to rely on.)

I have a possible solution for correcting out-of-order visibility in here for pending overrides, but handling the interactions with squash is still an open question (to me, at least.)

Reply:
Sorry, my wires were crossed here the other day and I wasn't thinking about this or communicating clearly: I was thinking about if we could reliably treat the data in the pending table as FIFO queue/log based on the ordering of the primary key, which is a different topic than transaction visibility. I'm not sure why I was mixing that up and combining the two topics. They are sort of related, but not really the same.

The data in the pending table should represent a consistent view of the state of overrides due to the transaction, and we shouldn't be able to end up in an inconsistent state where we see the side effects of a transaction prior to some other transaction that preceded it. We'd wouldn't be able to see A <- B without also seeing B <- C in the same set, regardless of their ordering within that set.

It is theoretically possible that the primary key of the records created within those transactions _don't_ appear in the order that they were committed — but based on the ordering of statements within the merge transaction and the locks that will be involved with those statements, I don't think this is an issue in practice either, B <- C should have a lower primary key than A <- B because of the row locks involved (at least those involved with person update and deletion, maybe others) and those statements preceding the insert to the pending overrides table which generates the sequence value.

For squash: it seems like it'd be ideal if we could keep the overrides writer running while squashes are occurring? I'm not sure how long squashes will take (though hopefully we'll find out next week) but it seems like that could delay the visible side effects of merges for a little bit? I guess it's not the end of the world if that's that case, overrides will still happen eventually, they'll just take longer to observe. I think there are things we could to do make it go smoother if the squash process is slow enough to be problematic, but probably better to actually focus on getting it running first. 🙂

Reply:
squash: yes we should be able to. 
The interesting case is: If the overrides writer updates a row after squash started processing that row ... if we remove only the rows that we processed (e.g. the problem comes, when some row got updated, e.g. we handled B <- A and now we have  C <- A (because afterwards overrides writer added C <- B). Now squash can still update all the events to have B instead of A, it would just not remove the C <- A row (might also be safe to remove, but I'm not sure, so would be easier not to). We can always keep extra rows in PG there without any risk.